Sources/SecureEnclave.swift

```diff
@@ -0,0 +1,181 @@
+        /// Data was retrieved from the keychain.
+        case success(Type)
+        /// User dismissed the user-presence prompt.
+        case userCancelled
+        /// No data was found for the requested key.
+        case itemNotFound
+        
+        // MARK: Equatable
+        
+        public static func ==(lhs: Result<Type>, rhs: Result<Type>) -> Bool {
+            switch lhs {
+            case let .success(lhsResult):
+                if case let .success(rhsResult) = rhs, lhsResult == rhsResult {
+                    return true
+                } else {
+                    return false
+                }
+            case .userCancelled:
+                if case .userCancelled = rhs {
+                    return true
+                } else {
+                    return false
+                }
+            case .itemNotFound:
+                if case .itemNotFound = rhs {
+                    return true
+                } else {
+                    return false
+                }
+            }
+        }
+    }
+    
+    // MARK: Internal Methods
+    
+    /// - parameter service: The service of the keychain slice we want to check if we can access.
+    /// - parameter identifier: A non-empty identifier that scopes the slice of keychain we want to access.
+    /// - returns: `true` if the keychain is accessible for reading and writing, `false` otherwise.
+    /// - note: Determined by writing a value to the keychain and then reading it back out.
+    internal static func canAccessKeychain(with service: Service, identifier: Identifier) -> Bool {
+        // To avoid prompting the user for Touch ID or passcode, create a Valet with our identifier and accessibility and ask it if it can access the keychain.
+        let noPromptValet: Valet
+        switch service {
+        case .standard:
+            noPromptValet = .valet(with: identifier, accessibility: .whenPasscodeSetThisDeviceOnly)
+        case .sharedAccessGroup:
+            noPromptValet = .sharedAccessGroupValet(with: identifier, accessibility: .whenPasscodeSetThisDeviceOnly)
+        }
+        
+        return noPromptValet.canAccessKeychain()
+    }
+    
+    /// - parameter object: A Data value to be inserted into the keychain.
+    /// - parameter key: A Key that can be used to retrieve the `object` from the keychain.
+    /// - parameter options: A base query used to scope the calls in the keychain.
+    /// - returns: `false` if the keychain is not accessible.
+    @discardableResult
+    internal static func set(object: Data, forKey key: String, options: [String : AnyHashable]) -> Bool {
+        // Remove the key before trying to set it. This will prevent us from calling SecItemUpdate on an item stored on the Secure Enclave, which would cause iOS to prompt the user for authentication.
+        _ = Keychain.removeObject(forKey: key, options: options)
+        
+        switch Keychain.set(object: object, forKey: key, options: options) {
```

Comment:
I _always_ explicitly iterate over every case using a switch rather than using `==`. I like the compile time safety switching over every case provides; if at some point in the future we add a case, I'll need to determine the right thing to do here before the code compiles again. If I just use `==`, then I can modify this code flow without revisiting it.